### PR TITLE
datafusion: Add network config fields for Private Service Connect

### PR DESCRIPTION
**Pull request by marblejenka**
PR for https://github.com/hashicorp/terraform-provider-google/issues/18339

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
datafusion: added fields `connection_type` and `private_service_connect_config` to enable using Private Service Connect with `google_data_fusion_instance` resources
```

### CODE REVIEW

File: mmv1/products/compute/NetworkAttachment.yaml [hunk 1]
```diff
@@ -13,7 +13,7 @@
 
 --- !ruby/object:Api::Resource
 name: 'NetworkAttachment'
-min_version: beta
```
Comment (marblejenka) [inline mmv1/products/compute/NetworkAttachment.yaml:16]: if https://github.com/GoogleCloudPlatform/magic-modules/pull/10966 get merged, we can remove this line.

File: mmv1/products/datafusion/Instance.yaml [hunk 2]
```diff
@@ -237,16 +249,54 @@ properties:
         description: |
           The IP range in CIDR notation to use for the managed Data Fusion instance
           nodes. This range must not overlap with any other ranges used in the Data Fusion instance network.
-        required: true
+        required: false
```
Comment (marblejenka) [inline mmv1/products/datafusion/Instance.yaml:252]: no longer required for the case that PSC used.

File: mmv1/products/datafusion/Instance.yaml [hunk 3]
```diff
@@ -237,16 +249,54 @@ properties:
...
         immutable: true
       - !ruby/object:Api::Type::String
         name: 'network'
         description: |
           Name of the network in the project with which the tenant project
           will be peered for executing pipelines. In case of shared VPC where the network resides in another host
           project the network should specified in the form of projects/{host-project-id}/global/networks/{network}
-        required: true
+        required: false
```
Comment (marblejenka) [inline mmv1/products/datafusion/Instance.yaml:260]: no longer required for the case that PSC used.

File: mmv1/products/datafusion/Instance.yaml [hunk 4]
```diff
@@ -237,16 +249,54 @@ properties:
...
         immutable: true
+      - !ruby/object:Api::Type::Enum
+        name: 'connectionType'
+        description: |
+          Optional. Type of connection for establishing private IP connectivity between the Data Fusion customer project VPC and
+          the corresponding tenant project from a predefined list of available connection modes.
+          If this field is unspecified for a private instance, VPC peering is used.
+        values:
+          - :CONNECTION_TYPE_UNSPECIFIED
```
Comment (marblejenka) [inline mmv1/products/datafusion/Instance.yaml:269]: maybe we can remove `CONNECTION_TYPE_UNSPECIFIED`.

File: mmv1/templates/terraform/examples/data_fusion_instance_psc.tf.erb [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+resource "google_compute_subnetwork" "psc" {
+  name   = "<%= ctx[:vars]["subnet_name"] %>"
+  region = "us-central1"
+
+  network       = google_compute_network.psc.id
+  ip_cidr_range = "10.0.0.0/16"
+}
+
+resource "google_compute_network_attachment" "psc" {
```
Comment (marblejenka) [inline mmv1/templates/terraform/examples/data_fusion_instance_psc.tf.erb:31]: `google_compute_network_attachment` is not in `google-provider` module and https://github.com/GoogleCloudPlatform/magic-modules/pull/10966 will solve the issue.

**Review by SarahFrench — COMMENTED** [review]
Hi @marblejenka - I've been prompted to review this PR but it looks like you're blocked waiting on this still? : https://github.com/GoogleCloudPlatform/magic-modules/pull/10969#discussion_r1639862235

Please re-request review when this PR is ready to proceed!

**Comment by marblejenka** [comment]
@SarahFrench Now it's ready for review.
There are some trial and error in the commit log because I thought network_config and private_service_connect_config could be changed. Please ignore them because the PATCH operation does not accept these changes.

File: mmv1/products/datafusion/Instance.yaml [hunk 6]
```diff
@@ -237,16 +249,59 @@ properties:
         description: |
           The IP range in CIDR notation to use for the managed Data Fusion instance
           nodes. This range must not overlap with any other ranges used in the Data Fusion instance network.
-        required: true
+        required: false
```
Comment (SarahFrench) [inline mmv1/products/datafusion/Instance.yaml:252]:
```suggestion
```

nit: Our convention is to omit the field if it should be false, as that's the default value when the field is unset.

File: mmv1/products/datafusion/Instance.yaml (hunk 4 above)
Comment (SarahFrench) [reply mmv1/products/datafusion/Instance.yaml:269]: Yeah, I'd recommend removing that value!

File: mmv1/products/datafusion/Instance.yaml (hunk 6 above)
Comment (SarahFrench) [reply mmv1/products/datafusion/Instance.yaml:252]: This suggestion is also true for other places in this PR, but I didn't want to create clutter by repeating it!

File: mmv1/products/datafusion/Instance.yaml [hunk 7]
```diff
@@ -237,16 +249,59 @@ properties:
...
+        description: |
+          Optional. Configuration for Private Service Connect.
+          This is required only when using connection type PRIVATE_SERVICE_CONNECT_INTERFACES.
+        required: false
         immutable: true
+        properties:
+          - !ruby/object:Api::Type::String
+            name: 'networkAttachment'
+            description: |
+              Required. The reference to the network attachment used to establish private connectivity.
```
Comment (SarahFrench) [inline mmv1/products/datafusion/Instance.yaml:285]: Currently the field's description says it's required but the field isn't marked as required. This means [the generated documentation describes the field as both required and optional](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-10969-old..auto-pr-10969#diff-980a1051f9f4b9f12e86b1fa3861a8e98253f6b2b1f60034ebf10bec0c51dd0fR382-R384). Could you please fix this discrepancy?

**Review by SarahFrench — COMMENTED** [review]
I've left some small comments, and currently I'm running the new acceptance test in our test environment (note to self, [here](https://hashicorp.teamcity.com/buildConfiguration/TerraformProviders_GoogleCloud_GOOGLE_BETA_MMUPSTREAMTESTS_GOOGLEBETA_PACKAGE_DATAFUSION/172289?buildTab=log&focusLine=1059&logView=linear&linesState=954)) and waiting to see the outcome there.

File: mmv1/products/datafusion/Instance.yaml (hunk 6 above)
Comment (marblejenka) [reply mmv1/products/datafusion/Instance.yaml]: I removed required: false

File: mmv1/products/datafusion/Instance.yaml (hunk 7 above)
Comment (marblejenka) [reply mmv1/products/datafusion/Instance.yaml]: I have modified the description

**Pull request merged**